src/python/pants/engine/target.py

```diff
@@ -1444,5 +1536,6 @@ def rules():
         find_valid_configurations,
         hydrate_sources,
         RootRule(TargetsToValidConfigurationsRequest),
+        RootRule(GenerateSourcesRequest),
```

Comment:
This should probably not be a `RootRule`: the conventions are nowhere near solid, but I think of them as being a bit like a public API, and in this case `GenerateSourcesRequest` is an implementation detail of `HydrateSourcesRequest`.

If a test wants to poke at those rules more directly, it can add a `RootRule` itself to do so. But I think that we should think of the roots that we expose in rulesets as their external inputs for general use.

Reply:
Hm, my understanding of RootRules is that you should use them whenever you directly inject a type into the graph, rather than deriving that type from other rules in the graph? Meaning, almost every `Request` class should be a `RootRule` because they are almost always directly created through a Python constructor, rather than being derived from some other rule. 

Is this mental model the wrong way of understanding `RootRule`?

Reply:
> Is this mental model the wrong way of understanding RootRule?

Yes, that is the wrong way to think about `RootRule`. See https://github.com/pantsbuild/pants/blob/master/src/python/pants/engine/README.md#gets-and-rootrules

In short: a Param enters the graph _either_ via a `Get` or via a `RootRule`. Things that enter as `Get`s do not need to be declared as roots. This is where the "root" in the name comes from: you only need to declare something a `RootRule` if it might come in at the "root" of a graph: ie, `scheduler.product_request`.

Reply:
Oh, huh. We're declaring way too many `RootRules` then, I think, in part from bad advice I gave Benjy. I'll clean that up.

cc @benjyw we shouldn't be using `RootRule` as much as I thought.